johnzon-jsonb/src/main/java/org/apache/johnzon/jsonb/JsonbPolymorphismHandler.java

```diff
@@ -0,0 +1,155 @@
+        }
+
+        return entries;
+    }
+
+    public Class<?> getTypeToDeserialize(JsonObject jsonObject, Class<?> clazz) {
+        validateJsonbTypeInfo(clazz);
+        validateOnlyOneParentWithTypeInfo(clazz);
+
+        JsonbTypeInfo typeInfo = Meta.getDirectAnnotation(clazz, JsonbTypeInfo.class);
+        if (typeInfo == null || !jsonObject.containsKey(typeInfo.key())) {
+            return clazz;
+        }
+
+        JsonValue typeValue = jsonObject.get(typeInfo.key());
+        if (!(typeValue instanceof JsonString)) {
+            throw new JsonbException("Property '" + typeInfo.key() + "' isn't a String, resolving "
+                    + "JsonbSubtype is impossible");
+        }
+
+        String typeValueString = ((JsonString) typeValue).getString();
+        for (JsonbSubtype subtype : typeInfo.value()) {
+            if (subtype.alias().equals(typeValueString)) {
+                return subtype.type();
+            }
+        }
+
+        throw new JsonbException("No JsonbSubtype found for alias '" + typeValueString + "' on " + clazz.getName());
+    }
+
+    /**
+     * Validates that only one parent class (superclass + interfaces) has {@link JsonbTypeInfo} annotation
+     *
+     * @param classToValidate class to validate
+     * @throws JsonbException validation failed
+     */
+    private void validateOnlyOneParentWithTypeInfo(Class<?> classToValidate) {
+        if (getParentClassesWithTypeInfo(classToValidate).size() > 1) {
+            throw new JsonbException("More than one interface/superclass of " + classToValidate.getName() +
+                    " has JsonbTypeInfo Annotation");
+        }
+    }
+
+    /**
+     * Validates {@link JsonbTypeInfo} on clazz.
+     * Validation fails either if any {@link JsonbSubtype#type()} is the same as clazz
+     * or if any clazz and {@link JsonbSubtype#type()} aren't compatible.
+     *
+     * @param classToValidate Class to validate
+     * @throws JsonbException validation failed
+     */
+    private void validateJsonbTypeInfo(Class<?> classToValidate) {
+        if (!classToValidate.isAnnotationPresent(JsonbTypeInfo.class)) {
+            return;
+        }
+
+        JsonbTypeInfo typeInfo = Meta.getDirectAnnotation(classToValidate, JsonbTypeInfo.class);
+        for (JsonbSubtype subtype : typeInfo.value()) {
+            if (!classToValidate.isAssignableFrom(subtype.type())) {
+                throw new JsonbException("JsonbSubtype '" + subtype.alias() + "'" +
+                        " (" + subtype.type().getName() + ") is not a subclass of " + classToValidate);
+            }
+        }
+    }
+
+    /**
+     * Collects all parent classes (superclass + interfaces) that have the JsonbTypeInfo annotation
+     * @param clazz base class
+     * @return List of classes with JsonbTypeInfo annotation
+     */
+    private List<Class<?>> getParentClassesWithTypeInfo(Class<?> clazz) {
```

Comment:
wonder if the following pattern isnt more relevant (pseudo code):

```
final var superClass = clazz.getSuperclass();
if (superClass != null && hasTypeInfo(superClass)) {
  return superClass;
}
return Stream.of(clazz.getInterfaces()).filter(it -> hasTypeInfo(it)).findFirst().orElse(null);
```

Overall idea is to avoid to create a list to keep only the first element. For interface case it will not change a lot but for parent case it will make it way more efficient. wdyt?